### PR TITLE
test: add more E2E tests (5/): scheduler watchers (1/)

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR adds more E2E tests, which verifies the behavior of the scheduler watchers.

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

N/A

### Special notes for your reviewer

N/A

### CODE REVIEW

**Comment by michaelawyu** [comment]
Per earlier agreements, this PR removes the provisional E2E test for member cluster label change response.

File: test/e2e/scheduler_watchers_test.go [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+
+				mcObj.Labels = nil
+				return hubClient.Update(ctx, mcObj)
+			}, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to drop the label from the cluster")
+		})
+
+		It("should keep the cluster in the scheduling decision (ignored during execution semantics)", func() {
+			targetClusterNames := []string{additionalClusterName1ForWatcherTests}
+			crpStatusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), targetClusterNames, nil)
+			Eventually(crpStatusUpdatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to update CRP status as expected")
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:336]: should use consistently?

File: test/e2e/scheduler_watchers_test.go [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+
+		It("can mark the new cluster as unhealthy", func() {
+			markMemberClusterAsUnhealthy(additionalClusterName1ForWatcherTests)
+		})
+
+		It("should keep the cluster in the scheduling decision", func() {
+			targetClusterNames := allMemberClusterNames
+			targetClusterNames = append(targetClusterNames, additionalClusterName1ForWatcherTests)
+			crpStatusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), targetClusterNames, nil)
+			Eventually(crpStatusUpdatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to update CRP status as expected")
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:403]: consistently?

File: test/e2e/scheduler_watchers_test.go [hunk 3]
```diff
@@ -0,0 +1,720 @@
...
+		})
+
+		It("can mark the new cluster as unhealthy", func() {
+			markMemberClusterAsUnhealthy(additionalClusterName1ForWatcherTests)
+		})
+
+		It("should keep the cluster as picked", func() {
+			targetClusterNames := []string{additionalClusterName1ForWatcherTests}
+			crpStatusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), targetClusterNames, nil)
+			Eventually(crpStatusUpdatedActual, consistentlyDuration, consistentlyInterval).Should(Succeed(), "Failed to update CRP status as expected")
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:712]: consistently?

File: test/e2e/utils_test.go [hunk 4]
```diff
@@ -32,25 +32,125 @@ import (
...
+		// Add a custom deletion blocker finalizer to the member cluster.
+		if err := hubClient.Get(ctx, types.NamespacedName{Name: name}, mcObj); err != nil {
+			return err
+		}
+
+		mcObj.Finalizers = append(mcObj.Finalizers, customDeletionBlockerFinalizer)
+		return hubClient.Update(ctx, mcObj)
+	}, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to mark member cluster as left")
+
+	Expect(client.IgnoreNotFound(hubClient.Delete(ctx, mcObj))).To(Succeed(), "Failed to delete member cluster")
```
Comment (zhiying-lin) [inline test/e2e/utils_test.go:146]: if we add the finalizer, it will never be 404?

File: test/e2e/scheduler_watchers_test.go [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+
+	clusterv1beta1 "go.goms.io/fleet/apis/cluster/v1beta1"
+	placementv1beta1 "go.goms.io/fleet/apis/placement/v1beta1"
+)
+
+const (
+	additionalClusterName1ForWatcherTests = "additional-cluster-1"
+	additionalClusterName2ForWatcherTests = "additional-cluster-2"
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:25]:
```suggestion
	fakeClusterName1ForWatcherTests = "additional-cluster-1"
	fakeClusterName2ForWatcherTests = "additional-cluster-2"
```
looks like they're fake cluster which does not install agents. Could you please rename them to call out it explicitly?

File: test/e2e/utils_test.go [hunk 6]
```diff
@@ -503,3 +598,84 @@ func ensureCRPAndRelatedResourcesDeletion(crpName string, memberClusters []*fram
 	// Delete the created resources.
 	cleanupWorkResources()
 }
+
+// verifyWorkPropagationAndMarkAsApplied verifies that works derived from a specific CPR have been created
+// for a specific cluster, and marks these works in the specific member cluster's
+// reserved namespace as applied.
+//
+// This is mostly used for simulating member agents for virtual clusters.
```
Comment (zhiying-lin) [inline test/e2e/utils_test.go:606]:
```suggestion
// This is mostly used for simulating member agents for fake clusters.
```
nit

File: test/e2e/scheduler_watchers_test.go (hunk 1 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go:343]: Done.

File: test/e2e/scheduler_watchers_test.go (hunk 3 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go]: Done.

File: test/e2e/scheduler_watchers_test.go (hunk 5 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go]: Done.

File: test/e2e/utils_test.go (hunk 6 above)
Comment (michaelawyu) [reply test/e2e/utils_test.go:606]: In actuality all of the clusters we use are fake

File: test/e2e/utils_test.go (hunk 4 above)
Comment (michaelawyu) [reply test/e2e/utils_test.go]: Done.

File: test/e2e/scheduler_watchers_test.go (hunk 2 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go:412]: Done.

File: test/e2e/scheduler_watchers_test.go [hunk 7]
```diff
@@ -331,14 +338,15 @@ var _ = Describe("responding to specific member cluster changes", func() {
 		})
 
 		It("should keep the cluster in the scheduling decision (ignored during execution semantics)", func() {
-			targetClusterNames := []string{additionalClusterName1ForWatcherTests}
+			targetClusterNames := []string{fakeClusterName1ForWatcherTests}
 			crpStatusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), targetClusterNames, nil)
 			Eventually(crpStatusUpdatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to update CRP status as expected")
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:343]: we don't need this eventually? cause in the line 324 & 325, you've already validated.

File: test/e2e/scheduler_watchers_test.go [hunk 8]
```diff
@@ -382,30 +390,32 @@ var _ = Describe("responding to specific member cluster changes", func() {
...
-			markMemberClusterAsUnhealthy(additionalClusterName1ForWatcherTests)
+			markMemberClusterAsUnhealthy(fakeClusterName1ForWatcherTests)
 		})
 
 		It("should keep the cluster in the scheduling decision", func() {
 			targetClusterNames := allMemberClusterNames
-			targetClusterNames = append(targetClusterNames, additionalClusterName1ForWatcherTests)
+			targetClusterNames = append(targetClusterNames, fakeClusterName1ForWatcherTests)
 			crpStatusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), targetClusterNames, nil)
 			Eventually(crpStatusUpdatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to update CRP status as expected")
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:412]: ditto

File: test/e2e/scheduler_watchers_test.go [hunk 9]
```diff
@@ -693,28 +711,30 @@ var _ = Describe("responding to specific member cluster changes", func() {
...
-			markMemberClusterAsUnhealthy(additionalClusterName1ForWatcherTests)
+			markMemberClusterAsUnhealthy(fakeClusterName1ForWatcherTests)
 		})
 
 		It("should keep the cluster as picked", func() {
-			targetClusterNames := []string{additionalClusterName1ForWatcherTests}
+			targetClusterNames := []string{fakeClusterName1ForWatcherTests}
 			crpStatusUpdatedActual := crpStatusUpdatedActual(workResourceIdentifiers(), targetClusterNames, nil)
-			Eventually(crpStatusUpdatedActual, consistentlyDuration, consistentlyInterval).Should(Succeed(), "Failed to update CRP status as expected")
+			Eventually(crpStatusUpdatedActual, eventuallyDuration, eventuallyInterval).Should(Succeed(), "Failed to update CRP status as expected")
```
Comment (zhiying-lin) [inline test/e2e/scheduler_watchers_test.go:731]: ditto

File: test/e2e/utils_test.go (hunk 6 above)
Comment (zhiying-lin) [reply test/e2e/utils_test.go:606]: ah, i remember we do create 3 kind clusters and install the agents there. is my understanding correct?

File: test/e2e/scheduler_watchers_test.go (hunk 7 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go:343]: Hence the `should keep the cluster in the scheduling decision` name

File: test/e2e/scheduler_watchers_test.go (hunk 8 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go:412]: Same as above.

File: test/e2e/scheduler_watchers_test.go (hunk 9 above)
Comment (michaelawyu) [reply test/e2e/scheduler_watchers_test.go:731]: Same as above.

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**